// ngx-uploadx/esm2020/lib/id.service.mjs
import { Injectable } from '@angular/core';
import { createHash } from './utils';
import * as i0 from "@angular/core";
export class IdService {
    generateId(uploader) {
        const print = JSON.stringify({
            ...uploader.metadata,
            type: uploader.constructor.name,
            endpoint: uploader.endpoint
        });
        return createHash(uploader.name + uploader.size).toString(16) + createHash(print).toString(16);
    }
}
IdService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IdService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
IdService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IdService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: IdService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });
