// ngx-uploadx/esm2020/lib/retry-handler.mjs
export var ErrorType;
(function (ErrorType) {
    ErrorType[ErrorType["NotFound"] = 0] = "NotFound";
    ErrorType[ErrorType["Auth"] = 1] = "Auth";
    ErrorType[ErrorType["Retryable"] = 2] = "Retryable";
    ErrorType[ErrorType["Fatal"] = 3] = "Fatal";
})(ErrorType || (ErrorType = {}));
const defaultRetryConfig = {
    maxAttempts: 8,
    shouldRestartCodes: [404, 410],
    authErrorCodes: [401],
    shouldRetryCodes: [408, 423, 429, 460],
    shouldRetry(code) {
        return code < 400 || code >= 500 || this.shouldRetryCodes.includes(code);
    },
    minDelay: 500,
    maxDelay: 50000,
    onBusyDelay: 1000,
    timeout: 0,
    keepPartial(statusCode) {
        return statusCode >= 400;
    }
};
/**
 * Retryable ErrorHandler
 */
export class RetryHandler {
    constructor(configOptions = {}) {
        this.attempts = 0;
        this.cancel = () => { };
        this.config = Object.assign({}, defaultRetryConfig, configOptions);
    }
    /**
     * Determine error type based on response code
     * @param code - HTTP response status code
     */
    kind(code) {
        this.attempts++;
        if (this.attempts > this.config.maxAttempts) {
            return ErrorType.Fatal;
        }
        if (this.config.authErrorCodes.includes(code)) {
            return ErrorType.Auth;
        }
        if (this.config.shouldRestartCodes.includes(code)) {
            return ErrorType.NotFound;
        }
        if (this.config.shouldRetry(code, this.attempts)) {
            return ErrorType.Retryable;
        }
        return ErrorType.Fatal;
    }
    /**
     * Wait before next retry attempt
     * @param time - Delay in ms
     */
    wait(time) {
        const ms = time || Math.min(2 ** (this.attempts - 1) * this.config.minDelay, this.config.maxDelay);
        const jitter = Math.floor(Math.random() * this.config.minDelay * this.attempts);
        let id;
        return new Promise(resolve => {
            this.cancel = () => {
                clearTimeout(id);
                resolve();
            };
            id = setTimeout(this.cancel, ms + jitter);
        });
    }
    /**
     * Observes value to reset retry attempts counter
     * @param value - Value to observe
     */
    observe(value) {
        this.observedValue !== value && (this.attempts = 0);
        this.observedValue = value;
    }
}
