// ngx-uploadx/esm2020/lib/uploadx.module.mjs
import { NgModule } from '@angular/core';
import { UPLOADX_OPTIONS } from './options';
import { UploadxDropDirective } from './uploadx-drop.directive';
import { UploadxDirective } from './uploadx.directive';
import * as i0 from "@angular/core";
export class UploadxModule {
    static withConfig(options) {
        return {
            ngModule: UploadxModule,
            providers: [{ provide: UPLOADX_OPTIONS, useValue: options }]
        };
    }
}
UploadxModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: UploadxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
UploadxModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.12", ngImport: i0, type: UploadxModule, imports: [UploadxDirective, UploadxDropDirective], exports: [UploadxDirective, UploadxDropDirective] });
UploadxModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: UploadxModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: UploadxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [UploadxDirective, UploadxDropDirective],
                    exports: [UploadxDirective, UploadxDropDirective]
                }]
        }] });
