// ngx-uploadx/esm2020/lib/uploadx.directive.mjs
import { EventEmitter, HostListener, Input, Output } from '@angular/core';
import { Directive } from '@angular/core';
import { takeWhile } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "./uploadx.service";
export class UploadxDirective {
    constructor(elementRef, renderer, uploadService) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.uploadService = uploadService;
        this.options = {};
        this.state = new EventEmitter();
    }
    set uploadx(value) {
        if (value) {
            this.options = value;
        }
    }
    set control(value) {
        if (value) {
            this.uploadService.control(value);
        }
    }
    ngOnInit() {
        const { multiple, allowedTypes } = { ...this.uploadService.options, ...this.options };
        multiple !== false && this.renderer.setAttribute(this.elementRef.nativeElement, 'multiple', '');
        allowedTypes &&
            this.renderer.setAttribute(this.elementRef.nativeElement, 'accept', allowedTypes);
        this.uploadService.events
            .pipe(takeWhile(() => this.state.observers.length > 0))
            .subscribe(this.state);
    }
    fileListener(files) {
        if (files?.length) {
            this.uploadService.handleFiles(files, this.options);
        }
    }
}
UploadxDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: UploadxDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }, { token: i1.UploadxService }], target: i0.ɵɵFactoryTarget.Directive });
UploadxDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.12", type: UploadxDirective, isStandalone: true, selector: "[uploadx]", inputs: { uploadx: "uploadx", options: "options", control: "control" }, outputs: { state: "state" }, host: { listeners: { "change": "fileListener($event.target.files)" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: UploadxDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[uploadx]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i1.UploadxService }]; }, propDecorators: { uploadx: [{
                type: Input
            }], options: [{
                type: Input
            }], control: [{
                type: Input
            }], state: [{
                type: Output
            }], fileListener: [{
                type: HostListener,
                args: ['change', ['$event.target.files']]
            }] } });
