// ngx-uploadx/esm2020/lib/uploadx-drop.directive.mjs
import { ContentChild, Directive, HostBinding, HostListener } from '@angular/core';
import { UploadxDirective } from './uploadx.directive';
import * as i0 from "@angular/core";
import * as i1 from "./uploadx.service";
export class UploadxDropDirective {
    constructor(uploadService) {
        this.uploadService = uploadService;
        this.active = false;
    }
    dropHandler(event) {
        this._stopEvents(event);
        this.active = false;
        const files = this.getFiles(event);
        if (files.length) {
            this.fileInput ? this.fileInput.fileListener(files) : this.uploadService.handleFiles(files);
        }
    }
    onDragOver(event) {
        this._stopEvents(event);
        if (event.dataTransfer?.items[0]?.kind === 'file') {
            if (this.fileInput?.options.multiple === false && event.dataTransfer.items.length > 1) {
                event.dataTransfer.dropEffect = 'none';
            }
            else {
                event.dataTransfer.dropEffect = 'copy';
                this.active = true;
            }
        }
    }
    onDragLeave(event) {
        this._stopEvents(event);
        this.active = false;
    }
    /**
     * Extracts the files from a `DragEvent` object
     */
    getFiles(event) {
        const dataTransfer = new DataTransfer();
        const items = event.dataTransfer?.items;
        if (items?.length) {
            for (let i = 0; i < items.length; i++) {
                const item = items[i];
                if (item.kind === 'file' && !item.webkitGetAsEntry()?.isDirectory) {
                    const file = item.getAsFile();
                    file && dataTransfer.items.add(file);
                }
            }
        }
        return dataTransfer.files;
    }
    _stopEvents(event) {
        event.stopPropagation();
        event.preventDefault();
    }
}
UploadxDropDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: UploadxDropDirective, deps: [{ token: i1.UploadxService }], target: i0.ɵɵFactoryTarget.Directive });
UploadxDropDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.12", type: UploadxDropDirective, isStandalone: true, selector: "[uploadxDrop]", host: { listeners: { "drop": "dropHandler($event)", "dragover": "onDragOver($event)", "dragleave": "onDragLeave($event)" }, properties: { "class.uploadx-drop-active": "this.active" } }, queries: [{ propertyName: "fileInput", first: true, predicate: UploadxDirective, descendants: true }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.12", ngImport: i0, type: UploadxDropDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[uploadxDrop]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.UploadxService }]; }, propDecorators: { active: [{
                type: HostBinding,
                args: ['class.uploadx-drop-active']
            }], fileInput: [{
                type: ContentChild,
                args: [UploadxDirective, { static: false }]
            }], dropHandler: [{
                type: HostListener,
                args: ['drop', ['$event']]
            }], onDragOver: [{
                type: HostListener,
                args: ['dragover', ['$event']]
            }], onDragLeave: [{
                type: HostListener,
                args: ['dragleave', ['$event']]
            }] } });
